src/com/amazon/ionelement/api/Element.kt

```diff
@@ -0,0 +1,257 @@
+package com.amazon.ionelement.api
+
+import com.amazon.ion.Decimal
+import com.amazon.ion.IntegerSize
+import com.amazon.ion.IonWriter
+import com.amazon.ion.Timestamp
+import java.math.BigInteger
+
+// QUESTION/TODO:  Verify that we are consistent about using *either* the Kotlin type name or the Ion type name.
+// i.e. Ion uses "float" file Kotlin's equivalent is "Double".  Int vs Long, etc.  I think in some places
+// we use one or the other but the intent is that the only place we use the Ion type name is for members of ElementType.
+// everywhere else should refer to the Kotlin type name.
+// The question here is: is this a good idea or if we should just use the Ion type name everywhere?
+
+/**
+ * TODO:  probably length documentation.
+ *
+ * TODO: when writing documentation, note that every implementation of [Element] must also implement [IonElement].
+ * The reason:  has to do with the need for all constructor functions accept [Element] but for the collection elements
+ * to return instances of [IonElement].
+ *
+ */
+interface Element {
+    /** The Ion data type of the current node.  */
+    val type: ElementType
+
+    /** This element's Ion metadata. */
+    val metas: MetaContainer
+
+    /** This element's Ion type annotations. */
+    val annotations: List<String>
+
+    /** Returns true if the current value is `null.null` or any typed null. */
+    val isNull: Boolean
+
+    /** Returns a copy of the current node with the specified additional annotations. */
+    fun withAnnotations(vararg additionalAnnotations: String): IonElement
+
+    /** Returns a copy of the current node with the specified additional annotations. */
+    fun withAnnotations(additionalAnnotations: Iterable<String>): IonElement =
+        withAnnotations(*additionalAnnotations.toList().toTypedArray())
+
+    /** Returns a copy of the current node without any annotations.  (Not recursive.) */
+    fun withoutAnnotations(): IonElement
+
+    /**
+     * Returns a copy of the current node with the specified additional metadata, overwriting any metas
+     * that exist with the same keys.
+     */
+    fun withMetas(additionalMetas: MetaContainer): IonElement
+
+    /**
+     * Returns a copy of the current node with the specified additional meta, overwriting any meta
+     * that previously existed with the same key.
+     *
+     * When adding multiple metas, consider [withMetas] instead.
+     */
+    fun withMeta(key: String, value: Any): IonElement = withMetas(metaContainerOf(key to value))
+
+    /** Returns a copy of the current node without any metadata.  (Not recursive.) */
+    fun withoutMetas(): IonElement
```

Comment:
I am replacing these functions on this interface with a single function:

```
    fun copy(annotations: List<String> = this.annotations, metas: MetaContainer = this.metas): IonElement
```

And adding them back as extension functions on `Element`.